src/test/java/io/confluent/examples/streams/CustomJoinWithTableTriggeringStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,365 @@
+        final private Duration maxWaitTimePerRecordForTableSideData = Duration.ofSeconds(5);
+        final private Duration frequencyToCheckForExpiredWaitTimes = Duration.ofSeconds(2);
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public void init(final ProcessorContext context) {
+          streamBufferStore = (KeyValueStore<String, Pair<Double, Instant>>) context.getStateStore(streamBufferStoreName);
+          tableStore = (KeyValueStore<String, Long>) context.getStateStore(tableStoreName);
+          this.context = context;
+          // Note: In practice, you will probably want to use `PunctuationType.STREAM_TIME`.  However, using stream time
+          // would make the test/validation setup in this example more complicated, hence we use wall clock time.
+          this.context.schedule(frequencyToCheckForExpiredWaitTimes, PunctuationType.WALL_CLOCK_TIME, this::punctuate);
+        }
+
+        @Override
+        public KeyValue<String, Pair<Double, Long>> transform(final String key, final Double value) {
+          return sendFullJoinRecordOrWaitForTableSide(key, value);
+        }
+
+        private KeyValue<String, Pair<Double, Long>> sendFullJoinRecordOrWaitForTableSide(final String key, final Double value) {
+          Long tableValue = tableStore.get(key);
+          if (tableValue != null) {
+            // We have data for both the stream and the table, so we can send a fully populated join message downstream
+            // immediately.
+            return KeyValue.pair(key, new Pair<>(value, tableValue));
+          } else {
+            // Don't send a join output just yet because we're still lacking table-side information.  Instead, buffer
+            // the current stream-side record, hoping that the table side will eventually see a matching record within
+            // `maxWaitTimePerRecordForTableSideData`.
+            streamBufferStore.put(key, new Pair<>(value, Instant.now()));
+            return null;
+          }
+        }
+
+        private void punctuate(long timestamp) {
+          sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(timestamp);
+        }
+
+        private void sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(long timestamp) {
+          try (KeyValueIterator<String, Pair<Double, Instant>> iterator = streamBufferStore.all()) {
+            while (iterator.hasNext()) {
+              KeyValue<String, Pair<Double, Instant>> record = iterator.next();
+              Duration delta = Duration.between(record.value.y, Instant.ofEpochMilli(timestamp));
+              if (delta.compareTo(maxWaitTimePerRecordForTableSideData) > 0) {
+                // Final attempt to fetch table-side data; we use that data even if it is null (indicates: missing).
+                Long tableValue = tableStore.get(record.key);
+                context.forward(record.key, new Pair<>(record.value.x, tableValue));
+                streamBufferStore.delete(record.key);
```

Comment:
Question: Does it matter whether the store deletion happens before or after `context.forward()`?

Reply:
If the downstream processor node also access this store, then yes it matters; otherwise, no.

Reply:
Ok. I reordered the statements so that the store deletion happens *before* the context.forward().

Reply:
It also matters for correctness (it's actually a know issue in our DSL implementation, too). If you delete first, and forward, and fail before forwarding finished, you loose data. At-least-once processing actually requires, to forward first and delete afterwards.

Reply:
Ack, reordered again so that forwarding happens BEFORE store deletion.